cirq-core/cirq/protocols/apply_unitary_protocol.py

```diff
@@ -70,10 +70,18 @@ class ApplyUnitaryArgs:
             dtype as the target tensor.
         axes: Which axes the unitary effect is being applied to (e.g. the
             qubits that the gate is operating on).
+        slices: Which slices the unitary effect is being applied to (e.g. the
+            subdimensions of qubits that the gate is operating on). By default
+            it applies to subspace 0..d-1, where d is the dimension of the
+            unitary effect.
     """
 
     def __init__(
-        self, target_tensor: np.ndarray, available_buffer: np.ndarray, axes: Iterable[int]
+        self,
+        target_tensor: np.ndarray,
+        available_buffer: np.ndarray,
+        axes: Iterable[int],
+        slices: Optional[Iterable[slice]] = None,
```

Comment:
Requiring the user to pass in a slice here is weird. I think better to let the user pass in the subspaces explicitly and then calculate the slice representation, throwing an error if none exists. This also makes it forward compatible if we ever are able to handle completely arbitrary subspaces.